server/src/main/java/io/crate/expression/reference/DocRefResolver.java

```diff
@@ -115,10 +119,31 @@ public DocRefResolver(List<ColumnIdent> partitionedByColumns) {
                         if (value instanceof List<?> list && list.stream().allMatch(Objects::isNull) &&
                             !(ref.valueType() instanceof ArrayType<?>)) {
                             return null;
+                        } else if (value instanceof Map<?, ?> map) {
+                            return replaceArraysWithNull(map, ref.valueType());
                         }
                         throw e;
                     }
                 });
         }
     }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private static Map<?, ?> replaceArraysWithNull(Map<?, ?> valueMap, DataType<?> valueType) {
+        Map newMap = new HashMap<>(valueMap.size());
+        for (var entry : valueMap.entrySet()) {
+            Object entryKey = entry.getKey();
+            Object entryValue = entry.getValue();
+            DataType<?> innerType = ((ObjectType) valueType).innerType((String) entryKey);
+            if (entryValue instanceof Map<?, ?> map) {
+                newMap.put(entryKey, replaceArraysWithNull(map, innerType));
+            } else if (entryValue instanceof List<?> list && list.stream().allMatch(Objects::isNull) &&
```

Comment:
This is a bit of a duplication of the Line 119. Can't make this method more general to consume any value and thus replace both if branches at Line 119? And decide if the value is returned by a nullable return value?